File: node/examples/pull-callback-cancellation.ts
import { createRPCClient, rpcCallbacks } from '../src/index.js';

interface DataService {
  pullForever(callbacks: { onChunk: (data: { batch: number; index: number }) => void }): AsyncGenerator<void>;
}

class DataServiceImpl {
  public teardownCount = 0;
  public producedBatches = 0;

  async *pullForever(callbacks: { onChunk: (data: { batch: number; index: number }) => void }) {
    console.log('pullForever started');
    try {
      for (let b = 0; ; b++) {
        for (let i = 0; i < 3; i++) {
          callbacks.onChunk({ batch: b, index: i });
        }
        this.producedBatches = b + 1;
        console.log(`Batch ${b} yielded`);
        yield;
      }
    } finally {
      this.teardownCount++;
      console.log(`Handler finally{} ran (teardownCount=${this.teardownCount})`);
    }
  }
}

async function main() {
  console.log('Pull-Callback Cancellation Example\n');

  const server = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'pull-callback-cancel-server',
  });
  await server.connect();

  const impl = new DataServiceImpl();
  const unsubHandler = await server.registerHandler('data', impl, { withoutDecorators: true });

  const client = createRPCClient({
    servers: ['nats://localhost:4222'],
    auth: { user: 'server', password: 'server_password' },
    name: 'pull-callback-cancel-client',
  });
  await client.connect();

  const service = client.createProxy<DataService>('data');

  let chunksBeforeBreak = 0;
  let chunksAfterBreak = 0;
  let didBreak = false;

  const cbs = rpcCallbacks(
    {
      onChunk: (_data: { batch: number; index: number }) => {
        if (didBreak) chunksAfterBreak++;
        else chunksBeforeBreak++;
      },
    },
    { oneway: ['onChunk'] },
  );

  console.log('Starting infinite generator, breaking after 3 batches...\n');

  let batchesConsumed = 0;
  for await (const _ of service.pullForever(cbs)) {
    batchesConsumed++;
    console.log(`Batch ${batchesConsumed - 1} consumed`);
    if (batchesConsumed >= 3) {
      console.log('break!');
      didBreak = true;
      break;
    }
  }

  // Wait briefly for any straggler callback messages.
  await new Promise((resolve) => setTimeout(resolve, 300));

  console.log('\nResults');
  console.log(`  Batches consumed on client:   ${batchesConsumed} (expected 3)`);
  console.log(`  Batches produced on server:   ${impl.producedBatches} (expected 3 or 4, +1 tolerated for in-flight)`);
  console.log(`  Chunks received before break: ${chunksBeforeBreak}`);
  console.log(`  Chunks received after break:  ${chunksAfterBreak} (expected 0)`);
  console.log(`  Handler teardown count:       ${impl.teardownCount} (expected 1)`);

  const ok = batchesConsumed === 3 && impl.teardownCount === 1 && impl.producedBatches <= 4 && chunksAfterBreak === 0;

  console.log(`\n${ok ? 'PASS' : 'FAIL'} — cancellation cleanly stops server + callbacks`);

  await unsubHandler();
  await client.disconnect();
  await server.disconnect();
}

main().catch(console.error);

File: python/examples/pull_callback_cancellation.py
#!/usr/bin/env python3
import asyncio
from collections.abc import AsyncGenerator
from typing import Any

from camera_ui_rpc import create_rpc_client, rpc_callbacks
from camera_ui_rpc.handler import CallbackInvoker


class DataService:
    def __init__(self) -> None:
        self.teardown_count = 0
        self.produced_batches = 0

    async def pull_forever(self, invoker: CallbackInvoker) -> AsyncGenerator[None, None]:
        print("pull_forever started")
        try:
            b = 0
            while True:
                if not invoker.active:
                    print(f"invoker inactive at batch {b} - exit loop")
                    return
                for i in range(3):
                    await invoker.invoke("onChunk", {"batch": b, "index": i})
                self.produced_batches = b + 1
                print(f"Batch {b} yielded")
                yield
                b += 1
        finally:
            self.teardown_count += 1
            print(f"teardown ran (count={self.teardown_count})")


async def main() -> None:
    print("Pull-Callback Cancellation Example (Python)\n")

    server = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "pull-callback-cancel-server",
        }
    )
    await server.connect()

    svc = DataService()
    unsub_handler = await server.register_handler("data", svc, without_decorators=True)

    client = create_rpc_client(
        {
            "servers": ["nats://localhost:4222"],
            "auth": {"user": "server", "password": "server_password"},
            "name": "pull-callback-cancel-client",
        }
    )
    await client.connect()

    proxy = client.create_proxy("data")

    chunks_before_cancel = 0
    chunks_after_cancel = 0
    did_cancel = False

    def on_chunk(_data: Any) -> None:
        nonlocal chunks_before_cancel, chunks_after_cancel
        if did_cancel:
            chunks_after_cancel += 1
        else:
            chunks_before_cancel += 1

    cbs = rpc_callbacks(onChunk=on_chunk, oneway=["onChunk"])

    print("Starting infinite generator, breaking after 3 batches...\n")

    batches_consumed = 0
    async for _ in proxy.pull_forever(cbs):
        batches_consumed += 1
        print(f"Batch {batches_consumed - 1} consumed")
        if batches_consumed >= 3:
            print("break!")
            did_cancel = True
            break

    # Allow teardown + any in-flight callback messages to settle.
    await asyncio.sleep(0.5)

    print("\nResults")
    print(f"  Batches consumed on client:   {batches_consumed} (expected 3)")
    print(
        f"  Batches produced on server:   {svc.produced_batches} (expected 3 or 4, tolerance for in-flight)"
    )
    print(f"  Chunks received before cancel: {chunks_before_cancel}")
    print(f"  Chunks received after cancel:  {chunks_after_cancel} (expected small, async drain)")
    print(f"  Server teardown count:         {svc.teardown_count} (expected 1)")

    ok = (
        batches_consumed == 3
        and svc.teardown_count == 1
        and svc.produced_batches <= 5
        and chunks_after_cancel <= 6
    )

    print()
    if ok:
        print("PASS - cancellation cleanly stops server + callbacks")
    else:
        print("FAIL - cancellation did not clean up properly")

    await unsub_handler()
    await client.disconnect()
    await server.disconnect()


if __name__ == "__main__":
    asyncio.run(main())
